File: src/ios/TwilioVideoPlugin.m
/********* TwilioVideo.m Cordova Plugin Implementation *******/

@import TwilioVideo;
#import <Cordova/CDV.h>
#import "TwilioVideoViewController.h"

@interface TwilioVideoPlugin : CDVPlugin <TwilioVideoViewControllerDelegate>
@property (nonatomic, copy) NSString* callbackId;
@end


@implementation TwilioVideoPlugin

- (void)open:(CDVInvokedUrlCommand*)command {
    NSString* room = command.arguments[0];
    NSString* token = command.arguments[1];
    NSString* remoteParticipantName = command.arguments[2];
    
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIStoryboard *sb = [UIStoryboard storyboardWithName:@"TwilioVideo" bundle:nil];
        TwilioVideoViewController *vc = [sb instantiateViewControllerWithIdentifier:@"TwilioVideoViewController"];
        self.callbackId = command.callbackId;
        vc.delegate = self;
        vc.accessToken = token;
        vc.remoteParticipantName = remoteParticipantName;
       // UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:vc];
      //  [vc.navigationItem setRightBarButtonItem:[[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(dismissTwilioVideoController)]];
         
        
        [self.viewController presentViewController:vc animated:YES completion:^{
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"opened"];
            [pluginResult setKeepCallbackAsBool:YES];
            [vc connectToRoom:room];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:self.callbackId];
        }];
    });

}

- (void)getTwilioVersion:(CDVInvokedUrlCommand*)command {
  CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:[TwilioVideo version]];
  [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) dismissTwilioVideoController {
    [self.viewController dismissViewControllerAnimated: YES completion: ^ {
        if (self.callbackId != nil) {
            NSString * cbid = self.callbackId;
            self.callbackId = nil;
            CDVPluginResult * pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK  messageAsString:@"closed"];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:cbid];
        }
    }];
}

-(void) dismiss {
    [self dismissTwilioVideoController];
}

-(void) onConnected:(NSString *)participantId participantSid:(NSString *)participantSid
{
  NSDictionary *dict = @{
	@"event": @"onConnected",
	@"participantId": participantId,
	@"participantSid": participantSid,
  };
  CDVPluginResult * pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK  messageAsDictionary:dict];
  [pluginResult setKeepCallbackAsBool:YES];
  [self.commandDelegate sendPluginResult:pluginResult callbackId:self.callbackId];
}

-(void) onDisconnected:(NSString *)participantId participantSid:(NSString *)participantSid
{
  NSDictionary *dict = @{
	@"event": @"onDisconnected",
	@"participantId": participantId,
	@"participantSid": participantSid,
  };
  CDVPluginResult * pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK  messageAsDictionary:dict];
  [pluginResult setKeepCallbackAsBool:YES];
  [self.commandDelegate sendPluginResult:pluginResult callbackId:self.callbackId];
}

@end

File: src/ios/TwilioVideoPlugin.swift
//
//  TwilioVideoPlugin.swift
//
// Copyright (C) 2019 by Cape Productions, Inc.
//
// Converted from TwilioVideoPlugin.m

import TwilioVideo

@objc(TwilioVideoPlugin) class TwilioVideoPlugin: CDVPlugin, TwilioVideoViewControllerDelegate {
    var callbackId: String?

    func open(_ command: CDVInvokedUrlCommand?) {
        let room = command?.arguments[0] as? String
        let token = command?.arguments[1] as? String
        let remoteParticipantName = command?.arguments[2] as? String

        DispatchQueue.main.async {
            let sb = UIStoryboard(name: "TwilioVideo", bundle: nil)
            if let vc = sb.instantiateViewController(withIdentifier: "TwilioVideoViewController") as? TwilioVideoViewController {
                self.callbackId = command?.callbackId
                vc.delegate = self
                vc.accessToken = token ?? ""
                vc.remoteParticipantName = remoteParticipantName ?? ""
                self.viewController.present(vc, animated: true) {
                    vc.connect(toRoom: room)
                    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "opened")
                    pluginResult?.keepCallback = true
                    self.commandDelegate.send(pluginResult, callbackId: self.callbackId)
                }
            }
        }

    }

    func getTwilioVersion(_ command: CDVInvokedUrlCommand?) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: TwilioVideo.version())
        commandDelegate.send(pluginResult, callbackId: command?.callbackId)
    }

    func getTwilioStats(_ command: CDVInvokedUrlCommand?) {
        TwilioMonitor.getStats { stats in
            if let stats = stats {
                print("Stats: \(stats)")
            }
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: stats)
            self.commandDelegate.send(pluginResult, callbackId: command?.callbackId)
        }
    }

    func dismissTwilioVideoController() {
        viewController.dismiss(animated: true) {
            if self.callbackId != nil {
                let cbid = self.callbackId
                self.callbackId = nil
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "closed")
                self.commandDelegate.send(pluginResult, callbackId: cbid)
            }
        }
    }

    func dismiss() {
        dismissTwilioVideoController()
    }

    func onConnected(_ participantId: String?, participantSid: String?) {
        let dict = [
            "event": "onConnected",
            "participantId": participantId,
            "participantSid": participantSid
        ]
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: dict as [AnyHashable : Any])
        pluginResult?.keepCallback = true
        commandDelegate.send(pluginResult, callbackId: callbackId)
    }

    func onDisconnected(_ participantId: String?, participantSid: String?) {
        let dict = [
            "event": "onDisconnected",
            "participantId": participantId,
            "participantSid": participantSid
        ]
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: dict as [AnyHashable : Any])
        pluginResult?.keepCallback = true
        commandDelegate.send(pluginResult, callbackId: callbackId)
    }
}
